Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -2614,8 +2616,8 @@ protected class ClassWriter : IClassWriter {
     }
 
     protected override void EmitDestructor(string source, Formal dtor, int formalNonGhostIndex, DatatypeCtor ctor, List<Type> typeArgs, Type bvType, ConcreteSyntaxTree wr) {
-      var dtorName = FormalName(dtor, formalNonGhostIndex);
-      wr.Write("(({0}){1}{2}).{3}", DtCtorName(ctor, typeArgs, wr), source, ctor.EnclosingDatatype is CoDatatypeDecl ? "._Get()" : "", dtorName);
+      var dtorName = dtor.HasName ? dtor.CompileName : ctor.CompileName + FormalName(dtor, formalNonGhostIndex);
```

Comment:
The expression `dtor.HasName ? dtor.CompileName : ctor.CompileName + FormalName(dtor, formalNonGhostIndex)` seems the same as what's on line 652, and it also has to be the same for correctness. Could you extract that into a method and use it in both locations?

Could you add a comment explaining why destructors for named formals don't need the `ctor.CompileName` qualification? Maybe refer to the formal name collision detection in https://github.com/dafny-lang/dafny/blob/master/Source/Dafny/Resolver.cs#L2164

Reply:
I think I would prefer defining `dtorName` without using `name` anyways, like so:

```
dtorName = arg.HasName ? arg.CompileName : ctor.CompileName + FormalName(arg, index)
```

So then sharing the expression with `EmitDestructor` becomes more attractive.